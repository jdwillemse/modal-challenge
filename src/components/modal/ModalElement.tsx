import {
  ReactElement,
  ReactNode,
  SyntheticEvent,
  useCallback,
  useEffect,
  useRef,
} from "react";

import css from "./modal.module.css";
import { useModalStore } from "../../stores/modalStore";

export interface ModalProps {
  showModal: boolean;
  children: ReactNode;
}

function ModalElement({ showModal, children }: ModalProps) {
  const dialogRef = useRef<HTMLDialogElement>(null);
  const { closeModal } = useModalStore();

  // close the modal when clicking outside of the content area
  const handleDialogClick = useCallback(
    (event: SyntheticEvent<HTMLElement>) => {
      if (event.target === dialogRef.current) {
        closeModal();
      }
    },
    [closeModal]
  );

  // setting open=true on the dialog element is not the preferred way to trigger a modal so the
  // opening and closing is handled here
  useEffect(() => {
    const ref = dialogRef.current;
    if (ref && showModal) {
      ref.showModal();
    }
    return () => {
      ref?.close();
    };
  }, [showModal]);

  return (
    <dialog
      onCancel={closeModal}
      className={css.modal}
      ref={dialogRef}
      aria-modal="true"
      aria-labelledby="modalTitle"
      onClick={handleDialogClick}
    >
      <div className={css.wrapper}>{children}</div>
    </dialog>
  );
}

ModalElement.Header = ({ children }: { children: ReactElement }) => (
  <div className={css.header}>
    <div className={css.titleWrapper} id="modalTitle">
      {children}
    </div>
  </div>
);

ModalElement.Body = ({
  children,
}: {
  children: ReactElement | ReactElement[];
}) => <div className={css.body}>{children}</div>;

ModalElement.Footer = ({
  children,
}: {
  children: ReactElement | ReactElement[];
}) => <div className={css.footer}>{children}</div>;

export default ModalElement;
