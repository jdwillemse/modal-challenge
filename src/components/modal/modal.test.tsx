import { describe, expect, test } from "vitest";
import { render, screen } from "@testing-library/react";

import ModalElement from "./ModalElement";

describe("modal", () => {
  beforeAll(() => {
    // these functions need to be mocked because jsdom does not currently support HTMLDialogElement
    HTMLDialogElement.prototype.showModal = vi.fn(function mock(
      this: HTMLDialogElement
    ) {
      this.open = true;
    });
    HTMLDialogElement.prototype.close = vi.fn(function mock(
      this: HTMLDialogElement
    ) {
      this.open = false;
    });
  });

  test("should not render when not shown", () => {
    render(<ModalElement showModal={false}>Nothing to see here</ModalElement>, {
      container: document.body,
    });
    expect(screen.queryByRole("dialog")).toBeNull();
  });

  test("should render with simple structure", async () => {
    render(<ModalElement showModal={true}>Simple modal</ModalElement>, {
      container: document.body,
    });
    expect(screen.queryByRole("dialog")).toMatchSnapshot();
  });

  test("should render with namespaced structure", () => {
    render(
      <ModalElement showModal={true}>
        <ModalElement.Header>
          <h2>Header</h2>
        </ModalElement.Header>
        <ModalElement.Body>
          <p>Body</p>
        </ModalElement.Body>
        <ModalElement.Footer>
          <span>Footer</span>
        </ModalElement.Footer>
      </ModalElement>,
      { container: document.body }
    );
    expect(screen.queryByRole("dialog")).toMatchSnapshot();
  });
});
