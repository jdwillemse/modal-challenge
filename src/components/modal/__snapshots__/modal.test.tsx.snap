// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`modal > should render with namespaced structure 1`] = `
<dialog
  aria-labelledby="modalTitle"
  aria-modal="true"
  class="modal"
  open=""
>
  <div
    class="wrapper"
  >
    <div
      class="header"
    >
      <div
        class="titleWrapper"
        id="modalTitle"
      >
        <h2>
          Header
        </h2>
      </div>
    </div>
    <div
      class="body"
    >
      <p>
        Body
      </p>
    </div>
    <div
      class="footer"
    >
      <span>
        Footer
      </span>
    </div>
  </div>
</dialog>
`;

exports[`modal > should render with simple structure 1`] = `
<dialog
  aria-labelledby="modalTitle"
  aria-modal="true"
  class="modal"
  open=""
>
  <div
    class="wrapper"
  >
    Simple modal
  </div>
</dialog>
`;
